Test `applyBlocks` with arbitrary sequences of transactions

jonathanknowles opened the pull request:
## Issue number

ADP-1783

## Summary

This PR:
- introduces generators and shrinkers for arbitrary sequences of transactions.
- uses transaction sequences to test various properties of the [`applyBlocks`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/src/Cardano/Wallet/Primitive/Model.hs#L339) function.

## Reviewing

It's recommended to review this PR commit by commit. Earlier commits introduce abstract types that are specialised to more concrete types in later commits.

## New property tests

- [`prop_applyBlocks_filteredTxs_{all,none,some}Ours`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs#L2653)
    - tests that `applyBlocks` returns all transactions that are relevant to the wallet and no others.
    - tests that `applyBlocks` returns transactions in precisely the same order that they appear on the blockchain.
- [`prop_applyBlocks_lastUTxO_{all,none,some}Ours`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs#L2710)
    - tests that `applyBlocks` applies the UTxO transition rules correctly w.r.t.:
        - the subset of transactions that are relevant to the wallet.
        - the subset of the global UTxO set belonging to the wallet.

## New types

- [`StateDeltaSeq state delta`](https://github.com/input-output-hk/cardano-wallet/pull/3483/commits/050b88ff3cbd8dd195e59208d103fe712695a607)
    - models a contiguous, abstract sequence of state transitions of the form:
      `state_0` → `delta_0_1` → `state_1` → `delta_1_2` → `state_2` → ... -> `state_n`
- [`TxSeq`](https://github.com/input-output-hk/cardano-wallet/pull/3483/commits/7d780a6a1bdd8a8f87337cd63dac09d31c1728e5)
    - specialises a `StateDeltaSeq` so that `state ~ UTxO` and `delta ~ Tx`.
    - maintains an invariant that for every transition satisfies the UTxO transition rules.
    - provides support for partitioning sequences into sub-sequences. 
- [`BlockSeq`](https://github.com/input-output-hk/cardano-wallet/pull/3483/commits/2ae205729e88d388fe5f322258dd7f0d1efc6ec8)
    - represents a sequence of blocks that starts at a given block height and slot number.
    - each block contains a sequence of transactions.

## Generating transaction sequences

See [`genTxSeq`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/src/Cardano/Wallet/Primitive/Types/TxSeq/Gen.hs#L91).

This PR generates transaction sequences according to the following simple algorithm:
1. Generate an initial sequence, consisting of a randomised initial UTxO and no transactions.
2. For `n` iterations:
   a. generate a random valid transaction `t` for the last UTxO in the sequence.
   b. apply transaction `t` to the last UTxO in the sequence, producing a updated sequence with an updated last UTxO.
   
## Shrinking transaction sequences

See [`shrinkTxSeq`](https://github.com/input-output-hk/cardano-wallet/blob/005cf008d999d4ee151af5cd480dbe669ab4f131/lib/core/src/Cardano/Wallet/Primitive/Types/TxSeq/Gen.hs#L120).

One of the biggest issues with testing properties involving sequences of transactions is that counterexamples can be very long and hard to interpret. A single transaction or UTxO set, when pretty printed, can easily fill half a screen. Long transaction sequences can easily fill many pages. Consequently, the pertinent part of a counterexample may be hidden within a huge sea of irrelevant details. In the event of a test failure, it's valuable to minimize the size of any counterexample so that just the pertinent information is presented.

When shrinking a `TxSeq`, it's important that all shrunken sequences maintain the UTxO transition rules. Therefore, individual transactions and UTxO sets cannot be shrunk in isolation: sequences must be shrunk as a whole.

There are many conceivable methods of shrinking a transaction sequence. This PR only provides support for some of these methods. Methods currently supported include:

 - shrinking to a prefix of the original sequence
 - shrinking to a suffix of the original sequence
 - removing block boundaries
 - removing individual assets
 - removing transactions that have no effect on the UTxO set
 - minimizing the set of assets that appear in UTxO sets and transactions
 - minimizing the set of transaction identifiers
 
The shrinker for `TxSeq` performs shrinking in phases, where each phase uses one of the above shrinking methods. Phases are ordered according to their aggressiveness (ability to shrink) and their efficiency (computation overhead): shrinking phases that are more aggressive and efficient are executed first.

jonathanknowles committed:
Add type `StateDeltaSeq`.

This commit adds the `StateDeltaSeq` type, along with generators,
shrinkers, and property tests.

The 'StateDeltaSeq' type provides a way to model an abstract sequence of
state transitions, where each transition consists of an initial state
value, a delta value, and a final state value.

jonathanknowles committed:
Add type `TxSeq`.

This commit adds the `TxSeq` type, along with generators, shrinkers, and
property tests.

The `TxSeq` type is a specialization of the `StateDeltaSeq` type,
with `UTxO` sets for states, and `Tx` values for deltas.

jonathanknowles committed:
Add type `NoneOurs`.

The `NoneOurs` type provides a convenient way to simulate a wallet that
owns no addresses or other entities.

As a result of this commit, we now have three convenient types for
simulating ownership:

- AllOurs:
  marks all entities as "ours" (belonging to a wallet)

- NoneOurs:
  marks all entities as "not ours" (not belonging to a wallet)

- IsOursIf:
  marks entities as "ours" if (and only if) some Boolean condition is
  satisfied.

jonathanknowles committed:
Add type `BlockSeq`.

A `BlockSeq` is sequence of blocks that starts at a given block height
and slot number.

jonathanknowles committed:
Add property tests for `applyBlocks`.

jonathanknowles committed:
Add function `labelInterval` to `QuickCheck.Extra`.

jonathanknowles committed:
Add function `isInjectiveOver` to `Data.Function.Utils`.

jonathanknowles committed:
Move `TxSeq` into new transaction module hierarchy.

A recent refactoring moved all transaction-related types into the
following module hierarchy:

Cardano.Wallet.Primitive.Types.Tx

This change moves the `TxSeq` module (and related modules) into this
hierarchy.

jonathanknowles committed:
Add function `isProperSubsetOf` to module `UTxO`.

This function is analogous to `Map.isProperSubmapOf`.

Before this change, we would have to write:
```
(someUTxO `UTxO.isSubSetOf` someOtherUTxO) && (someUTxO /= someOtherUTxO)
```

After this change, we can write:
```
someUTxO `UTxO.isProperSubsetOf` someOtherUTxO
```

jonathanknowles committed:
Add function `intersection` to module `UTxO`.

This function is analogous to `Map.intersection`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types/StateDeltaSeq.hs:
```diff
@@ -0,0 +1,541 @@
+data StateDeltaSeq state delta = StateDeltaSeq
+    { head :: !state
+    , tail :: !(Seq (delta, state))
+    }
```
It's probably not useful to use `!` here. (These bangs were left over from a previous version of the branch.)

According to the documentation for `Seq`:
> Despite sequences being structurally strict from a semantic standpoint, they are in fact implemented using laziness internally.

jonathanknowles replied:
However, there is also this:
https://hackage.haskell.org/package/strict-containers-0.1/docs/Data-Strict-Sequence.html

jonathanknowles replied:
However, it's currently difficult to experiment with changing this to [`Seq`](https://hackage.haskell.org/package/strict-containers-0.1/docs/Data-Strict-Sequence.html) from [`strict-containers`](https://hackage.haskell.org/package/strict-containers-0.1), as the version of [`cardano-base`](https://github.com/input-output-hk/cardano-base) on which we currently depend also exports a library called [`strict-containers`](https://github.com/input-output-hk/cardano-base/tree/7e3ddba98a61900181fe63cdd4c9ed9708d1a6a7/strict-containers), and Cabal (currently) doesn't make it possible to depend upon two libraries with identical names.

This naming conflict was fixed in https://github.com/input-output-hk/cardano-base/pull/308, but it'll take a while for this change to become available in our environment.

So for now, I'll shelve the experiment of changing this to the strict version, and remove the bang operators.

jonathanknowles replied:
Fixed in 8d5bcaecd8dae708d27ddb7aa45b77e6d570c825.

Unisay commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -151,6 +153,9 @@ balance =
 difference a b = a `excluding` Map.keysSet (unUTxO b)
 
+intersection :: UTxO -> UTxO -> UTxO
+intersection (UTxO a) (UTxO b) = UTxO (a `Map.intersection` b)
```
```suggestion
intersection (UTxO a) (UTxO b) = UTxO (Map.intersection a b)
```

jonathanknowles replied:
@Unisay I agree that it's nice to avoid mixing infix- and prefix-style here.

I've converted these to use prefix style in 4dafba07d0a37a397387cd363696372b90648bea.

Unisay commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO.hs:
```diff
@@ -166,6 +171,11 @@ isSubsetOf :: UTxO -> UTxO -> Bool
+-- | a ⊂ b
+isProperSubsetOf :: UTxO -> UTxO -> Bool
+isProperSubsetOf (UTxO a) (UTxO b) =
+    a `Map.isProperSubmapOf` b
```
```suggestion
isProperSubsetOf (UTxO a) (UTxO b) = a `Map.isProperSubmapOf` b
```

jonathanknowles replied:
I've converted this to use avoid infix application in 4dafba07d0a37a397387cd363696372b90648bea.

Unisay commented on lib/core/src/Data/Function/Utils.hs:
```diff
@@ -2,12 +2,33 @@
+--
+isInjectiveOver :: Ord b => (a -> b) -> Set a -> Bool
+isInjectiveOver f domain =
+    Set.size domain == Set.size (Set.map f domain)
```
```suggestion
isInjectiveOver f domain = Set.size domain == Set.size (Set.map f domain)
```

jonathanknowles replied:
Fixed in 4dafba07d0a37a397387cd363696372b90648bea.

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -325,6 +347,30 @@ spec = do
+                    & property
+            it "prop_applyBlocks_lastUTxO_noneOurs" $
+                prop_applyBlocks_lastUTxO_noneOurs
+                    & property
```
```suggestion
            it "prop_applyBlocks_filteredTxs_someOurs" $
                property prop_applyBlocks_filteredTxs_someOurs
            it "prop_applyBlocks_filteredTxs_allOurs" $
                property prop_applyBlocks_filteredTxs_allOurs
            it "prop_applyBlocks_filteredTxs_noneOurs" $
                property prop_applyBlocks_filteredTxs_noneOurs

        describe "lastUTxO" $ do
            it "prop_applyBlocks_lastUTxO_someOurs" $
                property prop_applyBlocks_lastUTxO_someOurs
            it "prop_applyBlocks_lastUTxO_allOurs" $
                property prop_applyBlocks_lastUTxO_allOurs
            it "prop_applyBlocks_lastUTxO_noneOurs" $
                property prop_applyBlocks_lastUTxO_noneOurs
```

jonathanknowles replied:
@Unisay Thanks for your suggestion!

I'd prefer to keep the current style of this block (and others like it), as:
- it does not violate our [Coding Standards](https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md);
- we already have a precedent for using this style in many other spec modules;
- it's currently easy to see the correspondence between property name and the string identifier (see https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r974847186);
- it's currently easy to insert/delete modifiers post hoc, for example:
    ```patch
                it "prop_applyBlocks_lastUTxO_someOurs" $
                    prop_applyBlocks_lastUTxO_someOurs
    +                   & verboseShrinking
    +                   & withMaxSuccess 1_000
    +                   & within 1_000_000
                        & property
    ```

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+    , initialSlotNo
+        :: SlotNo
+    , shrinkableTxSeq
+        :: ShrinkableTxSeq
```
```suggestion
    { initialBlockHeight :: Quantity "block" Word32
    , initialSlotNo :: SlotNo
    , shrinkableTxSeq :: ShrinkableTxSeq
```

jonathanknowles replied:
@Unisay I think your suggested style is also fine, though the existing style doesn't violate the Coding Standards (to my knowledge). This style is also used in quite a few of our existing modules. Perhaps this discussion could be held outside the context of this PR?

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+    blockSeqToBlockList blockSeq =
+        NE.fromList $ getZipList $ makeBlock
+            <$> ZipList (enumFrom $ blockSeq & initialBlockHeight)
+            <*> ZipList (enumFrom $ blockSeq & initialSlotNo)
```
```suggestion
            <$> ZipList (enumFrom $ initialBlockHeight blockSeq)
            <*> ZipList (enumFrom $ initialSlotNo blockSeq)
```

jonathanknowles replied:
I actually prefer the existing style here, as it emphasises the **_commonality_** between the two lines: in both cases we're enumerating from _attributes_ of `blockSeq`, and in both cases, we're enumerating from some `initial` value. If lenses were defined, we could have written:
```hs
            <$> ZipList (enumFrom $ blockSeq ^. #initialBlockHeight)
            <*> ZipList (enumFrom $ blockSeq ^. #initialSlotNo)
```

Unisay replied:
I see your point, usually I also prefer keeping similar parts on the lhs, and varying parts on rhs. 
In this particular case it leads to a mixture of `$` and `&` operators in the same expression and this makes it hard to parse in terms of associativity:
Example: `a $ b & c` - should it be associated to the left as `(a $ b) & c` or `a $ (b & c)` ?
Would you be willing to compromise on this:
```
            <$> ZipList (enumFrom (blockSeq & initialBlockHeight))
            <*> ZipList (enumFrom (blockSeq & initialSlotNo))
```
?

jonathanknowles replied:
Hi @Unisay 

I've included your suggestion in a78e4215baffe74e50d199e8a64c9772d9bcebb4.

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+            { header = BlockHeader
+                { blockHeight
+                , slotNo
+                , headerHash = Hash ""
```
I suggest to consider using something like `error "headerHash is irrelevant"` to fail fast and precise if assumption doesn't hold.

Unisay replied:
Actually, you already use it below, `shouldNotEvaluate "Block.headerHash"`

jonathanknowles replied:
@Unisay I did actually use `error` when I first wrote this function. 😄 

However, then I discovered that `headerHash` is a strict field of `BlockHeader`, and `applyBlocks` pattern matches on values of `BlockHeader` in order to construct values of `ChainPoint`. 

So using `error` here is not an option, unfortunately, unless we change `applyBlocks` itself.

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+        txSeq :: TxSeq
+        txSeq = blockSeqToTxSeq blockSeq
+
+        -- Makes a block using dummy values for fields that are not relevant
```
I am afraid that dummy values can mask a problem in case if some code suddenly relies on such a value.

jonathanknowles replied:
Hi @Unisay

I do understand the point you make here, and I think it's a good one, but it raises a question: what values should we use here instead?

Internally, the `applyBlocks` function does pattern match on these values, so we can't conveniently use `error`.

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+-- their fees.
+--
+nullifyFee :: Tx -> Tx
+nullifyFee = set #fee Nothing
```
IMO this function is below the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)

jonathanknowles replied:
> IMO this function is below the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold)

Fair enough!

I had originally intended to extend this function with knowledge of how `applyBlocks` modifies fees, but eventually didn't get around to that. After giving the matter some thought, I'd probably be in favour of modifying `applyBlocks` so that it doesn't modify the `fee` field at all. (This behaviour seems completely unnecessary.)

I've inlined the definition and removed the function in bf0b7364f59fcb6c24dc4c9ae2b7348577c4346d.

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+    -- TODO: we currently ignore transaction fees, as under some circumstances
+    -- the 'applyBlocks' function can modify fee values:
+    = fmap nullifyFee ourTxsReturned ====
+      fmap nullifyFee ourTxsExpected
```
Do we really want do deep-compare transactions or maybe just comparing their ids (hashes) suffice?

jonathanknowles replied:
> Do we really want do deep-compare transactions or maybe just comparing their ids (hashes) suffice?

Given the current type of `applyBlocks`, I think it's useful to verify that the function doesn't modify transactions in any meaningful way.

If `applyBlocks` were modified to return a set of transaction identifiers (i.e., a subset of the blocks supplied), or if we had a type-level proof that the function was incapable of modifying transactions, then we could drop this comparison entirely.

Unisay commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -533,6 +537,58 @@ shrinkMapWith shrinkKey shrinkValue
+            , "]"
+            ]
+  where
+    lo = (valueToCategorise `div` intervalSize) * intervalSize
+    hi = lo + (pred intervalSize)
```
```suggestion
    lo = valueToCategorise `div` intervalSize * intervalSize
    hi = lo + pred intervalSize
```

HeinrichApfelmus replied:
I would keep the parenthesis on the `lo` — the alternative parse

```
valueToCategorise `div` (intervalSize * intervalSize)
```

does not give the same result, and without parenthesis, one would have to look up the fixities of `div` and `*`.

jonathanknowles replied:
@Unisay As @HeinrichApfelmus suggested, I included the parentheses to aid readability, to spare readers the pain of looking up the relative precedences of these operators. At least with parentheses, there can be no doubt in the mind of the reader as to the desired precedence.

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/Tx/TxSeqSpec.hs:
```diff
@@ -0,0 +1,532 @@
+
+        describe "genTxSeq" $ do
+            it "prop_genTxSeq_isValid" $
+                prop_genTxSeq_isValid
```
I think I understand now what exactly you tried to achieve by putting `& property` in the end!
It makes a 1:1 alignment between the property function name and its test name. 
To me this looks like a job that is better done by machines, aka boilerplate. I'd either find a way to avoid having to repeat functions and their names or use the opportunity to provide a human-readable property description, e.g.:
```suggestion
            it "genTxSeq generates only valid transaction sequences" $
                prop_genTxSeq_isValid
```

jonathanknowles replied:
The main justification for repeating the property name in the description is that when a property fails, it's very easy to search for the property function name and locate it in the codebase. Unfortunately, `hspec` doesn't print the property name by default when there's a failure.

I actually don't think that these description fields are the right place to describe what the property does. Very often, there's no easy way to write a concise description that fits on one line. Putting it on multiple lines often requires the use of awkward line continuation characters.

I do think that if we need to provide a natural language description of the property, then we'd derive more benefit from putting that description in a **_comment_** attached to the property definition itself. Comments have the advantage of allowing many lines, and code examples.

If you think there are any properties that are currently unclear and would benefit from comments, I'd be happy to add them. Let me know!

In an ideal world, I agree with you that we could just generate these `hspec` structures automatically, instead of hand-coding them. Though that's arguably outside of the scope of this PR!

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs:
```diff
@@ -0,0 +1,654 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
```
@HeinrichApfelmus would be like
```suggestion
```
🤷🏼‍♂️

HeinrichApfelmus replied:
Yup! I would be like that. I'm in favor of [PatternSynonyms](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/pattern_synonyms.html), but `ViewPatterns` just mixes up the order of arguments in a confusing way without providing much benefit in terms of abstraction capabilities.

jonathanknowles replied:
@Unisay 

We do use `ViewPatterns` quite a bit throughout our codebase. (Around 15 modules already use it.)

Though I'd definitely be happy to adopt `PatternSynonyms` instead of `ViewPatterns` in future PRs!

Would you be happy with the following proposal:

Proposal:
1. Keep the current usage as it is (for this PR).
2. Create a follow-on PR that replaces all usages of `ViewPatterns` with `PatternSynonyms` (where feasible). That PR can also update other modules not included within the scope of this PR.

The proposal would unblock merging of this PR, and allow the replacement work to be prioritised independently. I'd be happy to do it as a gardening task. 👍🏻

HeinrichApfelmus replied:
> The proposal would unblock merging of this PR, and allow the replacement work to be prioritised independently. I'd be happy to do it as a gardening task. 👍🏻

As long as `ViewPatterns` goes away, I'm fine with that.  😂

jonathanknowles replied:
@HeinrichApfelmus wrote:
> As long as ViewPatterns goes away, I'm fine with that. 😂 

Okay 😄. I've created a ticket for this: ADP-2230. (@HeinrichApfelmus, I've tentatively added you as co-pilot for that ticket. Does that work for you?)

My proposal is:
- For this PR, we keep the current usage as it is.
- After this PR is merged, I will create a **follow-on** PR that addresses ADP-2230 for the whole wallet. There are only 15 modules; it shouldn't take too long to replace all the usages with `PatternSynonyms`, or something else (on a case-by-case basis).

@Unisay, are you also okay with this plan?

Unisay commented on lib/core/test/unit/Cardano/Wallet/Primitive/Types/StateDeltaSeqSpec.hs:
```diff
@@ -0,0 +1,654 @@
+        testLawsMany @StateDeltaSeq
+            [ bifoldableLaws
+            , bifunctorLaws
+            ]
```
Nice!

jonathanknowles replied:
Thanks! I'm very grateful for the existence of the `quickcheck-classes` library.

Unisay commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq/Gen.hs:
```diff
@@ -0,0 +1,342 @@
+
+boundedEnumSucc :: (Bounded a, Enum a, Ord a) => a -> Maybe a
+boundedEnumSucc a
+    | a >= maxBound = Nothing
```
I don't understand how a  `Bounded a` could be greater than its max bound? 
```suggestion
    | a == maxBound = Nothing
```

jonathanknowles replied:
> I don't understand how a Bounded a could be greater than its max bound?

In the general case, it's quite possible for someone to define `Bounded` and `Ord` instances for some type `t`, where there exists some value `a :: t` such that `a > maxBound`. It wouldn't be good practice to do this, but the compiler certainly wouldn't prevent it.

This successor function is only well-defined for values satisfying `< maxBound`. Therefore, it's not well-defined for values satisfying `>= maxBound`.

I'm not sure I see the advantage for limiting the boundary check to `==` here. An alternative to returning `Nothing` would be to raise a run-time error. But I cannot see a clear advantage, unless you can see one?

Unisay commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TxSeq.hs:
```diff
@@ -0,0 +1,544 @@
+-- This module provides the 'TxSeq' type and related functions.
+--
+-- The 'TxSeq' type provides a way to model a linear sequence of transactions
+-- and intermediate UTxO sets, where each individual transaction 't_i_j'' is
```
```suggestion
-- and intermediate UTxO sets, where each individual transaction 't_i_j' is
```

jonathanknowles replied:
You clearly have the eyes of an eagle! 🦅

Thanks!

jonathanknowles replied:
Fixed in 4dafba07d0a37a397387cd363696372b90648bea.

jonathanknowles committed:
Miscellaneous formatting fixes.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r972862847
https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r972863234
https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r972876832
https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r973179251
https://github.com/input-output-hk/cardano-wallet/pull/3483#issuecomment-1249174794

jonathanknowles committed:
Inline usages of `nullifyFee` and remove this function.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3483#pullrequestreview-1110546164

jonathanknowles committed:
Remove bang operators from definition of `StateDeltaSeq`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r972719881

jonathanknowles committed:
Use parentheses instead of `$` within `ModelSpec`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3483#discussion_r977510883
